cpp/log/frontend.cc

```diff
@@ -111,95 +105,40 @@ std::string Frontend::SubmitResultString(SubmitResult result) {
 }
 
 // static
-SubmitResult Frontend::GetSubmitError(
-    CertSubmissionHandler::SubmitResult result) {
+Status Frontend::GetSubmitError(CertSubmissionHandler::SubmitResult result) {
   CHECK_NE(result, CertSubmissionHandler::OK);
 
-  SubmitResult submit_result;
   switch (result) {
     case CertSubmissionHandler::EMPTY_SUBMISSION:
+      return util::Status(util::error::INVALID_ARGUMENT, "Empty submission.");
     case CertSubmissionHandler::INVALID_PEM_ENCODED_CHAIN:
-      submit_result = BAD_PEM_FORMAT;
-      break;
+      return util::Status(util::error::INVALID_ARGUMENT,
+                          "Invalid PEM encoded chain.");
     case CertSubmissionHandler::SUBMISSION_TOO_LONG:
-      submit_result = SUBMISSION_TOO_LONG;
-      break;
+      return util::Status(util::error::INVALID_ARGUMENT,
+                          "Submission too long.");
     case CertSubmissionHandler::INVALID_CERTIFICATE_CHAIN:
+      return util::Status(util::error::FAILED_PRECONDITION,
+                          "Invalid certificate chain.");
     case CertSubmissionHandler::UNKNOWN_ROOT:
-      submit_result = CERTIFICATE_VERIFY_ERROR;
-      break;
+      return util::Status(util::error::FAILED_PRECONDITION, "Unknown root.");
     case CertSubmissionHandler::PRECERT_CHAIN_NOT_WELL_FORMED:
-      submit_result = PRECERT_CHAIN_NOT_WELL_FORMED;
-      break;
+      return util::Status(util::error::FAILED_PRECONDITION,
+                          "Prechain not well formed.");
     case CertSubmissionHandler::INTERNAL_ERROR:
-      submit_result = INTERNAL_ERROR;
-      break;
+      return util::Status(util::error::INTERNAL, "Internal error.");
     default:
       LOG(FATAL) << "Unknown CertSubmissionHandler return code " << result;
   }
-  return submit_result;
 }
 
-void Frontend::UpdateStats(ct::LogEntryType type, SubmitResult result) {
-  if (type == ct::X509_ENTRY)
-    UpdateX509Stats(result);
-  else
-    UpdatePrecertStats(result);
-}
-
-void Frontend::UpdateX509Stats(SubmitResult result) {
-  lock_guard<mutex> lock(stats_mutex_);
-  switch (result) {
-    case ADDED:
-      ++stats_.x509_accepted;
-      break;
-    case DUPLICATE:
-      ++stats_.x509_duplicates;
-      break;
-    case BAD_PEM_FORMAT:
-      ++stats_.x509_bad_pem_certs;
-      break;
-    case SUBMISSION_TOO_LONG:
-      ++stats_.x509_too_long_certs;
-      break;
-    case CERTIFICATE_VERIFY_ERROR:
-      ++stats_.x509_verify_errors;
-      break;
-    case INTERNAL_ERROR:
-      ++stats_.internal_errors;
-      break;
-    case PRECERT_CHAIN_NOT_WELL_FORMED:
-      LOG(FATAL) << "invalid PRECERT_CHAIN_NOT_WELL_FORMED on an X509 certificate";
-    default:
-      LOG(FATAL) << "unknown SubmitResult enum value: " << result;
+void Frontend::UpdateStats(ct::LogEntryType type, const Status& status) {
```

Comment:
Just a suggestion: make this return the status, so you could use it inline above, like this:

    return UpdateStats(entry.type(), GetSubmitError(pre_result));